core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -898,12 +918,47 @@ public void testLockingBeforeAndAfterRelocated() throws Exception {
         // recovery can be now finalized
         recoveryThread.join();
         assertThat(shard.state(), equalTo(IndexShardState.RELOCATED));
-        try (Releasable ignored = shard.acquirePrimaryOperationLock()) {
+        try (Releasable ignored = acquirePrimaryOperationLockBlockingly(shard)) {
             // lock can again be acquired
             assertThat(shard.state(), equalTo(IndexShardState.RELOCATED));
         }
     }
 
+    public void testDelayedOperationsBeforeAndAfterRelocated() throws Exception {
+        assertAcked(client().admin().indices().prepareCreate("test").setSettings(
+            Settings.builder().put("index.number_of_shards", 1).put("index.number_of_replicas", 0)
+        ).get());
+        ensureGreen();
+        IndicesService indicesService = getInstanceFromNode(IndicesService.class);
+        IndexService test = indicesService.indexService(resolveIndex("test"));
+        final IndexShard shard = test.getShardOrNull(0);
+        assertBusy(() -> assertThat(shard.state(), equalTo(IndexShardState.STARTED)));
+        CountDownLatch latch = new CountDownLatch(1);
+        Thread recoveryThread = new Thread(() -> {
+            try {
+                latch.await();
+                shard.relocated("simulated recovery");
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        });
+
+        recoveryThread.start();
+        latch.countDown();
```

Comment:
it seems that the latch is useless? or maybe you wanted to enable the recovery half way through the "indexing operations"?  Also, this suffers from the same racing condition - if the ops that are now in flight when the relocation comes in are not first in the onLockAcquiredActions list, we will have a deadlock.